pkg/api/auth/public_api_handler.go

```diff
@@ -0,0 +1,129 @@
+package auth
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/rancher/harvester/pkg/config"
+	ctlcorev1 "github.com/rancher/wrangler-api/pkg/generated/controllers/core/v1"
+
+	dashboardauthapi "github.com/kubernetes/dashboard/src/app/backend/auth/api"
+	"github.com/pkg/errors"
+	"k8s.io/client-go/rest"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
+)
+
+const (
+	//action
+	actionQuery         = "action"
+	loginActionName     = "login"
+	skippableActionName = "skippable"
+	//default cluserName/userName/contextName
+	defaultRestConfigResourceName = "default"
+)
+
+func NewPublicAPIHandler(scaled *config.Scaled, restConfig *rest.Config) *PublicAPIHandler {
+	return &PublicAPIHandler{
+		secrets:      scaled.CoreFactory.Core().V1().Secret(),
+		restConfig:   restConfig,
+		tokenManager: scaled.TokenManager,
+	}
+}
+
+type PublicAPIHandler struct {
+	secrets      ctlcorev1.SecretClient
+	restConfig   *rest.Config
+	tokenManager dashboardauthapi.TokenManager
+}
+
+func (h *PublicAPIHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	action := strings.ToLower(r.URL.Query().Get(actionQuery))
+	if r.Method == http.MethodGet && action == skippableActionName {
+		rw.Header().Set("Content-type", "application/json")
+		rw.WriteHeader(http.StatusOK)
+		rw.Write(responseBody(SkippableResponse{AuthenticationSkippable: config.AuthenticationSkippable}))
+		return
+	}
+
+	if r.Method != http.MethodPost {
+		rw.WriteHeader(http.StatusMethodNotAllowed)
+		rw.Write(responseBody(TokenResponse{Errors: []string{"Only POST method is supported"}}))
+		return
+	}
+
+	if action != loginActionName {
+		rw.WriteHeader(http.StatusBadRequest)
+		rw.Write(responseBody(TokenResponse{Errors: []string{"Unsupported action"}}))
+		return
+	}
+
+	var input Login
+	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
+		rw.WriteHeader(http.StatusBadRequest)
+		rw.Write(responseBody(TokenResponse{Errors: []string{"Failed to decode request body, " + err.Error()}}))
+		return
+	}
+
+	tokenResp, statusCode, err := h.login(&input)
+	if err != nil {
+		rw.WriteHeader(statusCode)
+		rw.Write(responseBody(TokenResponse{Errors: []string{err.Error()}}))
+		return
+	}
+
+	rw.Header().Set("Content-type", "application/json")
+	rw.WriteHeader(http.StatusOK)
+	rw.Write(responseBody(tokenResp))
```

Comment:
Should also set cookie for response. And the cookie should be secure and http only.